enhancements/proposals/cleanup_after_scaling.md

```diff
@@ -0,0 +1,140 @@
+requiring a cleanup. Because this isn’t visible externally, the proposed solution won’t trigger a cleanup.
+ScyllaDB plans to automate cleanup and cover all the cases, until then we won’t cover these cases.
+
+## Design Details
+
+Scylla-operator will be extended with a controller responsible for running a cleanup procedure when needed.
+
+### Cleanup trigger
+
+The root cause of the cleanup requirement is that during cluster changes tokens are moved between nodes.
+There are two alternatives of detecting when this happened:
+
+1. ScyllaCluster topology changes - new rack is added/removed, number of members changes, or additional CPUs are
+   added/removed.
+2. Check tokens in Scylla API, detect when it changed.
+
+The first one is edge driven, there are multiple cases when tokens might move or not, for example if cpu resource change
+won’t be big enough to allocate additional core the cleanup is not needed. For hybrid clusters or multi datacenter
+clusters, topology change in the non-managed part should cause the cleanup of nodes in the managed part which is
+impossible to detect. It’s also harder to implement since StatefulSets, which changes need to be detected, doesn't allow
+for per node distinction, the state needs to be carried to other resources (for example member Service) which brings
+unnecessary coupling.
+
+The second one is checking the actual root cause, and seems more bullet proof.
+
+### Implementation
+
+Two new annotations in member services will be reconciled. Controller in the sidecar will
+reconcile `internal.scylla-operator.scylladb.com/current-token-ring-hash`
+and Operator will reconcile `internal.scylla-operator.scylladb.com/last-cleaned-up-token-ring-hash`.
+
+The `current` one will always reflect the hashed output of Scylla API `/storage_service/tokens` endpoint from a given
+node. The `last` will be either equal to `current` if it’s not yet present, or will be updated to current only when
+cleanup
+Job for a given node is `Completed`.
+
+The output of tokens endpoint is an json sorted array of strings representing numbers, the annotation will keep SHA512
+hash of it as a value.
+
+The cleanup call is synchronous, client only triggers the operation, and all the resource intensive operations are done
+by Scylla. The duration of the operation is tied to the size of the entire dataset and disk performance. Cleanup on node
+that don't require it, also impacts the disk IO.
+To prevent the controller sync loop from being too long, the cleanup controller in Operator will be responsible for
+managing cleanup Jobs. The preconditions for Job for given node are:
+
+* `current` and `last` token ring hash annotations values are different
+* All previous nodes have their `current` and `last` token ring hash annotations values equal - the order is rack order
+  in ScyllaCluster.spec and Pod ordinal - to prevent two cleanup Jobs running concurrently.
+* StatefulSet controller is not progressing - topology changes are done.
```

Comment:
should this be `Available=true,Progressing=False,Degraded=False`?

Reply:
No, because aggregated conditions is also computed from JobController conditions. Added what is checked.

Reply:
progressing==False doesn't mean all the nodes are UN, that's why I've suggested the other conditions as well.
Technically, this should also check generation for the status it sees.

There is also the question of whether we'd do a live call in case the informers are stale.

Reply:
`StatefulSetControllerProgressing=False`  means all required statefulsets are fully rolled out which means all reports UN.
https://github.com/scylladb/scylla-operator/blob/30f6e46908f00e9e493392b7570ad2dac0bb872e/pkg/controller/scyllacluster/sync_statefulsets.go#L597-L617
In addition it means upgrade is finished which is also thing we we would like to wait before cleanup.

Good point about generation, i'll cover it it implementation.

Reply:
> StatefulSetControllerProgressing=False means all required statefulsets are fully rolled out which means all reports UN.

I don't think that's what `Progressing` means. The controller currently sets `Progressing` condition because the algorithm "artificially" waits for a rollout internally but the implication goes only one way - if the controller waits for a rollout it sets the progressing condition to manifest it to external observers. It doesn't imply that `Progressing` condition means all nodes are UN.

`Progressing` - Is the controller waiting for something to happen before making progress?
`Available` - Is enough pods up/available? (Say for an HA service that can be 2 out of 3)
`Degraded` - Are some pods down? (Object can be both available and degrade if it's an HA service)

That's why our e2e also wait for those conditions and not just `Progressing`.

https://github.com/scylladb/scylla-operator/blob/9518217ba718d30af4343d85959c90b43a47c32a/test/e2e/utils/helpers.go#L148-L159

In general and as an example, say, in a fully rolled out cluster, a container gets suddenly restarted, that should only flip `Degraded` (possibly `Available` as well), not `Progressing` as the controller isn't taking any action or waiting for something before making a reconciliation action (unless there is one based on that).

Reply:
It means it's in the code at given revision, not what it is supposed to mean. Described expected behavior doesn't reflect what it is currently in the code.

But I agree relying on side effect isn't a good idea. Updated the proposal.